src/hammer-vlsi/hammer_vlsi/submit_command.py

```diff
@@ -0,0 +1,236 @@
+import atexit
+import subprocess
+import hammer_utils
+from typing import Callable, Iterable, List, NamedTuple, Optional, Dict, Any, Union
+from hammer_logging import HammerVLSIFileLogger, HammerVLSILogging, HammerVLSILoggingContext
+from hammer_config import HammerDatabase
+from functools import reduce
+
+class HammerSubmitCommand:
+
+    @abstractmethod
+    def submit(self, args: List[str], env: Dict[str,str], logger: HammerVLSILoggingContext, cwd: str = None) -> str:
+        """
+        Submit the job to the job submission system. This function MUST block until the command is complete.
+
+        :param args: Command-line to run; each item in the list is one token. The first token should be the command to run.
+        :param env: The environment variables to set for the command
+        :param logger: The logging context
+        :param cwd: Working directory (leave as None to use the current working directory).
+        :return: The command output
+        """
+        pass
+
+    @abstractmethod
+    def read_settings(self, settings: Dict[str, Any], tool_namespace: str) -> None:
+        """
+        Read the settings object (a Dict[str, Any]) into meaningful class variables
+
+        :param settings: A Dict[str, Any] comprising the settings for this command
+        :param tool_namespace: The namespace for the tool (useful for logging)
+        """
+        pass
+
+
+    @staticmethod
+    def get(tool_namespace: str, database: HammerDatabase) -> "HammerSubmitCommand":
+        """
+        Get a concrete instance of a HammerSubmitCommand for a tool
+
+        :param tool_namespace: The tool namespace to use when querying the HammerDatabase (e.g. "synthesis" or "par")
+        :param database: The HammerDatabase object with tool settings
+        """
+
+        submit_command_mode = database.get_setting(tool_namespace + ".submit.command", nullvalue="none")
+        # TODO This is sketchy
+        submit_command_settings = database.get_setting("vlsi.submit.settings", nullvalue=[]) + \
+            database.get_setting(tool_namespace + ".submit.settings", nullvalue=[]) # type: List[Dict[str, Dict[str, Any]]]
+
+        # Settings is a List[Dict[str, Dict[str, Any]]] object. The first Dict key is the submit command name.
+        # Its value is a Dict[str, Any] comprising the settings for that command.
+        # The top-level list elements are merged from 0 to the last index, with later indices overriding previous entries.
+        def combine_settings(settings: List[Dict[str, Dict[str, Any]]], key: str) -> Dict[str, Any]:
+            return reduce(hammer_utils.add_dicts, map(lambda d: d[key], settings))
+
+        submit_command = None # type: Optional[HammerSubmitCommand]
+        if submit_command_mode == "none" or submit_command_mode == "local":
+            # Do not read the options, return immediately
+            return HammerLocalSubmitCommand()
+        elif submit_command_mode == "lsf":
```

Comment:
Do you envision hammer library users being able to add custom submit commands? Right now I think this function hardcodes the list of available ones.

Reply:
Yes, it does. I do think we should have a way to monkey patch custom submit commands in, however.

Reply:
Opened #223